### DISCUSSION TITLE
SQLAlchemy not returning timezone aware timestamp with TIMESTAMP(timezone=True)

### DISCUSSION THREAD

**Discussion by shoang22** [sqlalchemy/sqlalchemy#11791] [Usage Questions]
I have the following Oracle table:

```python
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column
from sqlalchemy.dialects.oracle import TIMESTAMP
from sqlalchemy.sql import func


class Base(DeclarativeBase):
    pass

class Foo(Base):
    __tablename__ = "foo"

    id: Mapped[int] = mapped_column(Identity(), primary_key=True)
    last_modified = mapped_column(TIMESTAMP(timezone=True), default=func.now())
```

When I execute a select statement, I expect to get back a timezone-aware datetime object:

```python
async def test_get_date(session: AsyncSession):
    foo = Foo()
    session.add(foo)
    await session.commit()

    stmt = select(Foo)
    row = (await session.execute(stmt)).scalar()

    assert row.last_modified.tzinfo is not None
```

But the test fails with the following error:

```python
>       assert row.last_modified.tzinfo is not None
E       assert None is not None
E        +  where None = datetime.datetime(2024, 8, 26, 17, 14, 19, 851).tzinfo
E        +    where datetime.datetime(2024, 8, 26, 17, 14, 19, 851) = <arthur.database.models.Foo object at 0x10a174110>.last_modified

tests/test_db.py:58: AssertionError
====================================================================================== short test summary info =======================================================================================
FAILED tests/test_db.py::test_get_date - assert None is not None
```

Is there something I'm missing? Interestingly, when I set timezone=False and I pass in a timezone aware object, I do get a timezone aware object back when I query the database, but I would like to be able to get timezone aware objects back in cases where I let Oracle generate a default value.

**Comment by zzzeek** [comment]
hi -

these datetime objects come straight from the driver, either cx_oracle or oracledb.   SQLAlchemy does not currently intercept this data in any way and it's passed straight on.

I would recommend asking over on the OracleDB discussion tracker at https://github.com/oracle/python-oracledb/discussions how come the driver is not handling this automatically.

For cx_oracle / oracledb specifically, SQLAlchemy does have what's known as oracle "outputtypehandlers" in place that can re-interpret data as it comes from the driver.   to date we have not installed such a driver to attempt to produce an alternate datetime object, if this fell upon us it would be a SQLAlchemy 2.1 thing as it would be a major behavioral change.    We'd also need feedback from the above oracledb discussion tracker to know how to do this.

so first step is you should ask on the oracledb tracker.  

you can use this demonstration script to illustrate the problem without involving SQLAlchemy

```py
import oracledb
from datetime import datetime, timezone


conn = oracledb.connect(
    user="scott",
    password="tiger",
    dsn=oracledb.makedsn("oracle18c", 1521, service_name="xe"),
)


cursor = conn.cursor()

try:
    cursor.execute("drop table datetime_table")
except:
    pass

cursor.execute(
    """
CREATE TABLE datetime_table (
    dt TIMESTAMP,
    dttz TIMESTAMP WITH TIME ZONE
)
"""
)


dt = datetime.now().replace(microsecond=0)
dttz = datetime.now(timezone.utc).astimezone().replace(microsecond=0)


cursor.execute(
    "INSERT INTO datetime_table (dt, dttz) VALUES (:dt, :dttz)",
    {"dt": dt, "dttz": dttz},
)

cursor.execute("SELECT dt, dttz FROM datetime_table")
row = cursor.fetchone()

assert row[0] == dt, f"{row[0]} != {dt}"
assert row[1] == dttz, f"{row[1]} != {dttz}"


```

**Reply by zzzeek** [reply]
> , I do get a timezone aware object back when I query the database

this is likely the same value you placed in the first place that was still present in the ORM identity map and was not actually fetched from the database.

**Reply by shoang22** [reply]
> I would recommend asking over on the OracleDB discussion tracker at https://github.com/oracle/python-oracledb/discussions how come the driver is not handling this automatically.

Thanks for the sample and next steps. For reference, I opened a discussion [here](https://github.com/oracle/python-oracledb/discussions/389).
